repos/keg-components/src/components/textToggle/textToggle.js

```diff
@@ -0,0 +1,179 @@
+ * @param {Number=} props.collapsedHeight - optional. height of the textview when collapsed
+ * @param {Component=} props.CustomToggle - optional toggle component to use instead of the default
+ * @param {string} props.fadeColor - optional color for the fade on collapsed text
+ */
+export const TextToggle = props => {
+  const {
+    text,
+    styles,
+    isExpandedInit = false,
+    className,
+    CustomToggle,
+    onToggleChange,
+    togglePosition = 'right',
+    collapsedHeight = 100,
+    fadeColor = 'white',
+  } = props
+
+  if (!text) return null
+  const [ expanded, setExpanded ] = useState(isExpandedInit)
+
+  const styleHelper = useMemo(
+    () => ({
+      styles,
+      togglePosition,
+    }),
+    [ styles, togglePosition ]
+  )
+
+  const mainStyle = useStylesCallback(
+    buildStyles,
+    [ togglePosition, styles ],
+    styleHelper
+  )
+
+  const [ textMaxHeight, setTextMaxHeight ] = useState(0)
+
+  const showToggle = shouldDisplayToggler(collapsedHeight, textMaxHeight)
+
+  const onToggleCb = useCallback(() => {
+    setExpanded(!expanded)
+    isFunc(onToggleChange) && onToggleChange(!expanded)
+  }, [ expanded, onToggleChange ])
+
+  const onTextLayout = useCallback(
+    event => {
+      const height = event.nativeEvent.layout.height
+      if (textMaxHeight === height) return
+      setTextMaxHeight(height)
+    },
+    [ textMaxHeight, setTextMaxHeight ]
+  )
+
+  return (
+    <View
+      style={[mainStyle.main]}
+      className={useClassList('keg-text-toggle', className)}
+    >
+      <Drawer
+        collapsedHeight={collapsedHeight}
+        toggled={expanded}
+      >
+        <Text
+          style={mainStyle.text}
+          onLayout={onTextLayout}
+        >
+          { text }
+        </Text>
+      </Drawer>
+
+      { showToggle && !expanded && (
+        <LinearGradient
+          colors={[ 'rgba(255,255,255,0)', fadeColor ]}
+          style={mainStyle.linearGradient}
+        />
+      ) }
+      { showToggle && (
+        <ToggleComponent
+          onPress={onToggleCb}
+          isExpanded={expanded}
+          styles={mainStyle.toggleComponent}
+          CustomComponent={CustomToggle}
+        />
+      ) }
+    </View>
+  )
+}
+
+/**
+ * only show toggle if user-defined height DNE
+ *  or the given text does not fit in the user-defined height
+ * @param {number=} minHeight
+ * @param {number=} textMaxHeight
+ *
+ * @return {boolean}
+ */
+const shouldDisplayToggler = (minHeight, textMaxHeight) => {
+  return !minHeight || textMaxHeight > minHeight
+}
+
+/**
+ * ToggleComponent
+ * @param {object} props
+ * @param {Function} props.onPress - when the component is pressed
+ * @param {object} props.styles
+ * @param {Component=} props.CustomComponent - optional Component to override the default text component
+ * @param {boolean} props.isExpanded - whether to show expanded state or not
+ */
+const ToggleComponent = ({ onPress, styles, CustomComponent, isExpanded }) => {
```

Comment:
We should add props the override the default text. I may not want a custom component. I just want to change what the text is. Something like `closedToggleText` and `expandedToggleText`

Reply:
agreed. updated